### ISSUE TITLE
Update `Drawer` to support Material 3

### ISSUE THREAD

**Issue by darrenaustin** [flutter/flutter#103551]
As part of the [Material 3 specification](https://m3.material.io/components/icon-buttons/overview) there are updates to the Navigation drawer visuals:

<img width="411" alt="Screen Shot 2022-05-11 at 5 15 20 PM" src="https://user-images.githubusercontent.com/19588/167967461-23309f3a-2cd8-45d3-a3b8-1ff1b408433f.png">

Currently Flutter only has the `Drawer` widget, but it is pretty bare-bones. We should look into possibly creating a new NavigationDrawer. It would be good if that could be integrated with the `NavigationDestination` used by the `NavigationBar` as well.

**Comment by ArishSultan** [comment]
@rydmike @darrenaustin I think we will need to redesign the existing `Scaffold` or make a new Scaffold with material 3 capabilities since the current `Scaffold` is only capable of showing an elevated `Drawer`, since the drawer gets wrapped with `DrawerController` behind the scenes. there is no way to implement an always visible drawer using Theme only, the new Scaffold should also have support for `NavigationRail`.

Share your thoughts.

**Comment by bernaferrari** [comment]
@ArishSultan if you want to be radical, deprecate drawer and add `sideMenu` to the Scaffold, which can be a Drawer that gets automatically expanded/shrinked depending on the screen size, or a rail that becomes navigation on phone.

**Comment by mhoad** [comment]
Just to quickly speak to the point @bernaferrari mentioned, just went through that process of getting that to work and it was much more painful than I would have anticipated for such a fundamental but of functionality, it feels like there is a lot of room here to improve the DX and the idea they mentioned is ultimately what I wish I could have done in the first place.

**Comment by elbeicktalat** [comment]
> @darrenaustin even the current Drawer container widget has some challenges when it comes to being able to deliver a M3 themed version of it.

I don't think that is an issue caused by the actual `Drawer`, this issue is related to the `Scaffold`.

What I have tried:
- Use `Directionality` widget, but it doesn't works since it change the text direction also.

To solve this issue I suggest the following:
1. Add border radius property in the `Drawer`, this will be applied for the regular drawer.
2. Add a new property to the `Scaffold` to determine the border radius for the end drawer.

Thanks!

**Comment by elbeicktalat** [comment]
@rydmike  I don't agree with you, the drawer widget is a very simple widget so to me it doesn't make a sense having a new one.

Yes it's more easy but not the best solution to me.

**Comment by rydmike** [comment]
@elbeicktalat I actually agree with you, it was @bernaferrari that suggested making a new one. Still a new additional drawer/menu widget with more out of the box features would also be useful, but for different reasons. In that case I think the existing simple Drawer container should still exist and be available as starting point for own custom menus, like today.

My comment was only concerning making the current Drawer have more theming properties. Could be something like visible corner radius, and screen edge corner radius (in addition to current general border radius, well Shape actually). Then when those widget props or theme props are used, let the widget figure out which corners should be rounded based on current directionality or if it is used in Scaffold drawer or endDrawer property. The last one might be tricky, since the widget itself or theme for it has no knowledge of that.

TLDR: 
I just want to be able to Shape theme the Drawer border radius for its visible side, so correct side is rounded regardless of if Drawer is used in Scaffold drawer or endDrawer, and be directionality aware as well for these cases as well.

**Comment by elbeicktalat** [comment]
I don't know if we could integrate the `Navigation Destination` used by the `Navigation Bar` only in one PR. since there is a lot to do.

**Comment by elbeicktalat** [comment]
This requires #108162

**Comment by hangyujin** [comment]
I created a  doc to discuss about drawer/navigationDrawer in M3 , please feel free to comment!   

https://docs.google.com/document/d/1tlgCIN5VYGM8LBT08y29OyeLb9EIhg2HgSo5UgmhXfk/edit?usp=sharing&resourcekey=0-FN54My8BuIgqRHhpshaEfg

**Comment by MarcusWichelmann** [comment]
Is it possible to make the children of `NavigationDrawer` scrollable? I have more `NavigationDrawerDestination`s than fit on every screen.

As the implementation looks like, this is not possible?

**Comment by MarcusWichelmann** [comment]
I also whonder how safe areas should be correctly handled in regard to scroling? Maybe the widget should provide a good default behavior for this?

**Comment by MarcusWichelmann** [comment]
Now reported here: https://github.com/flutter/flutter/issues/116782

**Comment by darrenaustin** [comment]
This drawer issue is fixed with #115668. The scrolling issue will be tracked in #116782.
